### PR TITLE
FIX: wallet cards not updating after refresh

### PR DESCRIPTION
**Pull request by ncoelho**
## Summary

Wallet carousel cards stayed stale after balance/tx refresh: wallets are mutated in place, and `React.memo` on `WalletCarouselItem` skipped re-renders when the same object references were re-published.

Masked previously by `#8810` (removed `useExtendedNavigation` / storage-driven `onPress` churn) and `#8418` (moved update status off the cards).

### Fix (minimal)

1. **Drop `React.memo`** on `WalletCarouselItem` so in-place balance/tx changes show when the parent re-renders.
2. **Publish stays in `saveToDisk`** — both `refreshAllWalletTransactions` and `fetchAndSaveWalletTransactions` already end with `saveToDisk()` → `setWallets([...])`. One publish path; no duplicate pre-save `setWallets`.
3. **`extraData={wallets}`** on the home `SectionList` so the carousel row updates when wallets change.
4. **Share `walletHasPendingTransaction()`** for pending-pill logic.
5. **FlatList `extraData` after `{...props}`** so callers cannot clobber the update signal.

### Not in this PR

- No DrawerList sync duplication (existing effect already mirrors `wallets` from storage).
- No revision-string / `dataRevision` algebra.

## Test plan

- [x] `tests/unit/wallet-carousel-refresh.test.tsx` (~165 lines)
  - `walletHasPendingTransaction`: on-chain + Lightning in-flight vs failed (exported helper)
  - One parametrized carousel republish case: balance + pending + label/unit/hideBalance — FlatList animate on/off + drawer path
  - PureComponent row-cache test: why `SectionList extraData={wallets}` is required when section data is static (not a fake inline-`renderItem` rerender)
  - Caller `extraData` cannot clobber FlatList update signals
- [x] Unit file passes locally
- [ ] Manual: pull-to-refresh updates card balances
- [ ] Manual: snap refresh updates the visible card

<div><a href="https://cursor.com/agents/bc-01a038df-61e6-7aa0-8cde-2b0c213875f8?cursor_ref=pr_footer&cursor_cta=open_in_web"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-01a038df-61e6-7aa0-8cde-2b0c213875f8&cursor_ref=pr_footer&cursor_cta=open_in_cursor"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

File: tests/unit/wallets-carousel-data-revision.test.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    };
+    const before = getWalletCarouselItemDataRevision(wallet);
+
+    wallet.__setTransactions([{ confirmations: 0 }]);
+    const after = getWalletCarouselItemDataRevision(wallet);
+
+    assert.notStrictEqual(before, after);
+  });
+
+  it('stays stable when unrelated mutable fields are unchanged', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallets-carousel-data-revision.test.ts]: **[HIGH]** Calling the same function twice on an untouched mock and calling it a 'stability' test? Even my control-group potatoes mutate more than this. CONTRIBUTING says no bullshit tests — this one verifies determinism, not that unrelated fields were ignored.

File: components/WalletsCarousel.tsx [hunk 2]
```diff
@@ -64,6 +64,34 @@ const WALLET_CARD_TEXT_OPACITY = 0.85;
...
+ * on `WalletCarouselItem` only shallow-compares props, so without a primitive that changes
+ * when those fields change, cards keep showing stale balances after an update.
+ *
+ * Prefer passing this from `renderItem` (computed after mutation) rather than reading
+ * `item.getBalance()` inside a custom memo comparator — for in-place mutation, prev/next
+ * `item` are the same reference and would both report the new balance.
+ */
+export const getWalletCarouselItemDataRevision = (item: TWallet): string => {
+  const isLightningShaped = item.type === LightningCustodianWallet.type || item.type === LightningArkWallet.type;
+  const hasPendingTx = isLightningShaped
```
Comment (GladosBlueWallet) [inline components/WalletsCarousel.tsx]: **[HIGH]** Lightning pending uses `ispaid === false && !tx.failed`, but your tests only poke on-chain `confirmations`. Copy-paste the predicate twice, test it zero times — classic human maintenance strategy. Ark dead-swap edge case? Untested. Enjoy the stale 'pending' pill when these drift.

File: tests/unit/wallets-carousel-data-revision.test.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import assert from 'assert';
+
+import { getWalletCarouselItemDataRevision } from '../../components/WalletsCarousel';
+import { TWallet } from '../../class/wallets/types';
+
+const makeWallet = (overrides: Partial<{
```
Comment (GladosBlueWallet) [inline tests/unit/wallets-carousel-data-revision.test.ts]: **[HIGH]** Forty-three lines of mock factory to assert four string inequalities. CONTRIBUTING asked for tests, not a miniature wallet simulator. Bloated test scaffolding — my favorite kind of false confidence.

File: components/WalletsCarousel.tsx [hunk 4]
```diff
@@ -64,6 +64,34 @@ const WALLET_CARD_TEXT_OPACITY = 0.85;
 
 export const getWalletCarouselItemWidth = (screenWidth: number) => Math.round(screenWidth * 0.82 > 375 ? 375 : screenWidth * 0.82);
 
+/**
```
Comment (GladosBlueWallet) [inline components/WalletsCarousel.tsx]: **[MEDIUM]** Twelve lines of JSDoc explaining React.memo like it's the Aperture Science Employee Handbook. CONTRIBUTING: minimal change, readable code — not a dissertation on why props exist.

File: tests/unit/wallets-carousel-data-revision.test.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      state.transactions = transactions;
+    },
+  } as unknown as TWallet & {
+    __setBalance: (balance: number) => void;
+    __setLatestTransactionTime: (time: number) => void;
+    __setTransactions: (transactions: any[]) => void;
+  };
+};
+
+describe('getWalletCarouselItemDataRevision', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallets-carousel-data-revision.test.ts]: **[MEDIUM]** Revision string tracks label, hideBalance, and preferredBalanceUnit — none tested. Miss one field in a future edit and cards ghost-update again. Edge cases were requested; you brought half a checklist.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**FIX** for stale wallet cards after refresh — wallets mutate in place, `React.memo` shallow-compares, cards lie. The `dataRevision` snapshot is the right idea and should work when `saveToDisk` → `setWallets` fires. Still duplicated pending-tx logic waiting to diverge, and the new tests include one that proves absolutely nothing. Science remains imperfect. Humans remain predictable.

### Inline findings (could not anchor on diff)

- `components/WalletsCarousel.tsx:471` — **[HIGH]** Same pending-tx predicate, second copy, ten lines apart. Card text reads one ruleset; `dataRevision` reads another — until someone edits one. Minimal diff? No. Minimal trust? Also no.

File: components/WalletsCarousel.tsx [hunk 6]
```diff
@@ -908,11 +921,12 @@ const WalletsCarousel = forwardRef<CarouselListRefType, WalletsCarouselProps>((p
     },
   });
 
+  const dataRevisionKey = useMemo(() => data.map(wallet => (wallet ? getWalletCarouselItemDataRevision(wallet) : '')).join('||'), [data]);
```
Comment (GladosBlueWallet) [inline components/WalletsCarousel.tsx]: **[HIGH]** `dataRevisionKey` memoized on `[data]` — the one reference that never changes during in-place mutation. Stale on purpose. The comment at 953 pretends this forces FlatList updates; reality is a new `extraData` array literal when a parent deigns to re-render. Misleading dead weight. Classic over-engineering.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**Bug fix (wallet cards stale after refresh)** — replaces deleted `useExtendedNavigation` re-render churn with a `dataRevision` snapshot for `React.memo`/`FlatList`. Pull-to-refresh probably works (loading state toggles). Silent carousel snap refresh still won't repaint. You built a revision-string Rube Goldberg machine instead of calling `setWallets` after refresh. Adequate for the test chamber. Not for production humans.

### Inline findings (could not anchor on diff)

- `screen/wallets/WalletsList.tsx:277` — **[HIGH]** Silent snap refresh — no loading flag, no status churn. React 18 ignores `NONE→NONE`. Wallet mutates in place; your precious `dataRevision` never reaches the memo. Card stays stale. Congratulations on fixing refresh by not refreshing the UI.
- `components/Context/StorageProvider.tsx:407` — **[HIGH]** Refresh finishes here without `setWallets`. Same object references forever. That forced this entire carousel revision sideshow — and it still fails whenever nothing else re-renders the tree. One line at the source beats exported string algebra.
- `components/WalletsCarousel.tsx:305` — **[MEDIUM]** `React.memo` only sees `dataRevision` if `renderItem` runs again. No parent re-render after silent refresh means memo keeps serving the old card. The prop is a key under glass — useless without a re-render trigger you forgot to wire.

File: tests/unit/wallet-has-pending-transaction.test.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    getLabel: () => 'Wallet',
+    getPreferredBalanceUnit: () => 'BTC',
+    getTransactions() {
+      return this.transactions;
+    },
+  };
+  return w as MutableWallet;
+};
+
+describe('walletHasPendingTransaction', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-has-pending-transaction.test.ts]: **[HIGH]** You deleted the revision regression suite and replaced it with mock-wallet predicate trivia. The stale-card fix—no memo plus publish—has zero test coverage. Even I log failed experiments.

File: components/WalletsCarousel.tsx [hunk 8]
```diff
@@ -279,272 +289,252 @@ const iStyles = StyleSheet.create({
...
-      (toValue: number) => {
-        pressScale.value = withSpring(toValue, { damping: 13, stiffness: 180, mass: 0.9 });
-      },
-      [pressScale],
-    );
-
-    useEffect(() => {
-      dragScale.value = withSpring(isDraggingActive ? dragActiveScale : 1, { damping: 16, stiffness: 200, mass: 1 });
-    }, [isDraggingActive, dragActiveScale, dragScale]);
+export const WalletCarouselItem: React.FC<WalletCarouselItemProps> = ({
```
Comment (GladosBlueWallet) [inline components/WalletsCarousel.tsx:292]: **[HIGH]** Unwrapping `React.memo` somehow re-indented the entire component—~250 lines of diff noise. CONTRIBUTING asked for minimal diffs, not a whole-chamber reformat for science.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**Intent:** FIX — wallet carousel cards stayed stale after refresh because `WalletCarouselItem` was `React.memo`'d while balances mutate in-place on the same object references.

**Approach:** Drop `React.memo`, call `setWallets([...BlueApp.getWallets()])` during refresh, add `extraData={wallets}` on the wallets `SectionList`, extract `walletHasPendingTransaction`. Mechanism is sound for the home carousel path.

**Issues:** No regression test for the actual stale-card fix; ~250 lines of re-indent churn; drawer carousel still unsynced on balance-only updates. Commits follow CONTRIBUTING prefixes but five iterations on one fix is not atomic.

### Inline findings (could not anchor on diff)

- `screen/wallets/DrawerList.tsx:175` — **[MEDIUM]** When wallet count is unchanged you never dispatch `SetWallets`; drawer carousel keeps stale balances while WalletsList gets the fix. Half the subjects updated. How… selective.

File: screen/wallets/DrawerList.tsx [hunk 9]
```diff
@@ -173,7 +173,9 @@ const DrawerList: React.FC<DrawerContentComponentProps> = memo((props: DrawerCon
         dispatch({ type: WalletActionType.SetWallets, wallets });
       }, 600);
     } else {
+      // Balance/tx refresh publishes a new wallets array with the same IDs — keep the drawer in sync.
       prevWalletIds.current = wallets.map(wallet => wallet.getID());
+      dispatch({ type: WalletActionType.SetWallets, wallets });
```
Comment (GladosBlueWallet) [inline screen/wallets/DrawerList.tsx]: **[HIGH]** Redundant `SetWallets` dispatch — the effect at line 201 already syncs `state.wallets` on every `wallets` change. You built two identical plumbing runs for the same water. Bloat. The facility frowns.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    );
+    assert.strictEqual(
+      walletHasPendingTransaction(makeWallet({ type: LightningArkWallet.type, transactions: [{ ispaid: false, failed: true }] })),
+      false,
+    );
+  });
+});
+
+describe('WalletCarouselItem refresh publish', () => {
+  it('shows updated balance after in-place mutation when parent re-renders (setWallets publish)', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[HIGH]** Test claims 'setWallets publish' but never invokes StorageProvider, FlatList, or SectionList — just `rerender`s the component directly. You're verifying memo removal in a vacuum, not the actual stale-card failure mode. Performance art, not science.

File: components/Context/StorageProvider.tsx [hunk 11]
```diff
@@ -400,6 +400,9 @@ export const StorageProvider = ({ children }: { children: React.ReactNode }) =>
             ]);
 
             console.debug('[refreshAllWalletTransactions] Saving data to disk');
+            // Publish before/via save so UI updates even when the header status stays NONE
+            // (silent snap refresh) and wallet instances were mutated in place.
+            setWallets([...BlueApp.getWallets()]);
```
Comment (GladosBlueWallet) [inline components/Context/StorageProvider.tsx]: **[MEDIUM]** Pre-save `setWallets` then `saveToDisk` calls `setWallets` again at line 165 — two array publishes per refresh, two re-render waves. Intentional early UI, yes; still wasteful thermodynamics.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**Intent:** FIX — wallet carousel cards stayed stale after balance/tx refresh because `React.memo` blocked re-renders on in-place-mutated wallet objects (regression after #8810 removed `useExtendedNavigation`).

**Verdict:** Core diagnosis is sound — drop `React.memo`, publish wallets from `StorageProvider`, wire `extraData`. Not a catastrophe. But you duplicated sync logic the facility already had, left sibling refresh paths unfixed, and your new test is a direct-component rerun cosplaying as an integration test. Adequate. Disappointingly adequate.

### Inline findings (could not anchor on diff)

- `components/Context/StorageProvider.tsx:458` — **[MEDIUM]** `fetchAndSaveWalletTransactions` still mutates wallets in place and only publishes via `saveToDisk` — no pre-save `setWallets` like line 405. Single-wallet refresh from WalletDetails, push listeners, etc. can leave carousel cards stale until disk save finishes. You patched one airlock and called it done.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+      true,
+    );
+    assert.strictEqual(
+      walletHasPendingTransaction(makeWallet({ type: LightningArkWallet.type, transactions: [{ ispaid: false, failed: true }] })),
+      false,
+    );
+  });
+});
+
+describe('wallet carousel refresh publish', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx:94]: **[HIGH]** Your carousel refresh test renders `WalletsCarousel` in isolation. The actual fix also lands at `screen/wallets/WalletsList.tsx:560` — `extraData={wallets}` on a `SectionList` whose carousel section data is a static string. That virtualization path is the other half of the experiment. Untested. How… convenient.

Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx:94]: **[HIGH]** You only assert balance text (`1 BTC` → `2.5 BTC`). The stale-card bug also affects pending pills and latest-tx footer via `walletHasPendingTransaction` at `components/WalletsCarousel.tsx:444`. Zero coverage that those strings refresh after in-place mutation. One happy path. For science.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  return w as MutableWallet;
+};
+
+describe('walletHasPendingTransaction', () => {
+  it('detects pending on-chain txs via confirmations === 0', () => {
+    assert.strictEqual(walletHasPendingTransaction(makeWallet({ transactions: [{ confirmations: 1 }] })), false);
+    assert.strictEqual(walletHasPendingTransaction(makeWallet({ transactions: [{ confirmations: 0 }] })), true);
+  });
+
+  it('treats unpaid Lightning/Ark swaps as pending, but not failed ones', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[HIGH]** `walletHasPendingTransaction` at `components/WalletsCarousel.tsx:69` branches on both `LightningCustodianWallet.type` and `LightningArkWallet.type`. Your tests only feed Ark at line 78. Custodian wallets get no edge-case coverage — the predicate could rot on half its input and your tests would throw confetti anyway.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+  });
+});
+
+describe('wallet carousel refresh publish', () => {
+  it('updates card balance when wallets are re-published as a new array of the same mutated instances', () => {
+    // Mirrors StorageProvider.saveToDisk: setWallets([...BlueApp.getWallets()]) after in-place fetch.
+    const wallet = makeWallet({ balance: 100_000_000 });
+    const onPress = jest.fn();
+
+    const screen = render(<WalletsCarousel data={[wallet]} onPress={onPress} animateChanges={false} />);
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[MEDIUM]** Production runs `animateChanges={true}` (`screen/wallets/WalletsList.tsx:332`); your regression harness disables it. Reanimated entering/layout wrappers in `renderItem` (`components/WalletsCarousel.tsx:792-798`) are the path users actually walk. Testing the easy mode is not testing.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
Bug-fix PR: stale wallet carousel cards after refresh, caused by `React.memo` on `WalletCarouselItem` skipping re-renders when wallet objects mutate in-place. Fix removes memo, merges `extraData` after `{...props}`, adds `SectionList` `extraData={wallets}`. Core logic is coherent. The regression suite, however, is thinner than a test chamber observation window — it never exercises the `SectionList` wiring that is half the fix, skips pending-footer refresh, and ignores `LightningCustodianWallet` in the shared predicate tests.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+      true,
+    );
+    assert.strictEqual(
+      walletHasPendingTransaction(makeWallet({ type: LightningCustodianWallet.type, transactions: [{ ispaid: false, failed: true }] })),
+      false,
+    );
+  });
+});
+
+describe('wallet carousel refresh publish', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx:94]: **[HIGH]** Your regression chamber tests balance and pending, then declares victory. The deleted revision suite also covered label, hideBalance, and unit mutations — none of which you still assert after dropping `dataRevision`. CONTRIBUTING wanted edge cases; you kept the easy ones.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    wallet.balance = 250_000_000;
+    wallet.transactions = [{ confirmations: 0 }];
+    screen.rerender(<WalletsCarousel data={[wallet]} onPress={onPress} animateChanges={true} />);
+
+    assert.ok(screen.getByText('2.5 BTC'));
+    assert.ok(screen.getByText(loc.transactions.pending));
+    assert.strictEqual(screen.queryByText('1 BTC'), null);
+  });
+
+  it('updates carousel inside SectionList when extraData wallets array is republished', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[HIGH]** SectionList gets a test; the drawer carousel (`isFlatList={false}`, `screen/wallets/DrawerList.tsx:252`) does not. Same stale-card failure mode, different plumbing. One passing experiment does not validate the whole facility.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+      latestTransactionTime: Date.now(),
+      transactions: [{ confirmations: 1 }],
+    });
+    const onPress = jest.fn();
+
+    const screen = render(<WalletsCarousel data={[wallet]} onPress={onPress} animateChanges={true} />);
+    assert.ok(screen.getByText('1 BTC'));
+    assert.strictEqual(screen.queryByText(loc.transactions.pending), null);
+
+    wallet.balance = 250_000_000;
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx:108]: **[MEDIUM]** You mutate the wallet, then rerender with a brand-new `[wallet]` array every time. Production only updates when `saveToDisk` republishes; in-place mutation without that publish still leaves cards frozen — and your tests would never catch it.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**Bug fix PR** — wallet carousel cards stayed stale after pull-to-refresh because `React.memo` blocked in-place wallet mutations and `SectionList` never re-rendered the carousel row (static section data). Fix: drop memo, add `extraData={wallets}` on the list, harden FlatList `extraData` after `{...props}`. Mechanism is sound given `saveToDisk` always republishes a fresh array. Test suite covers the happy path and pending-pill helper but shed edge-case coverage when `dataRevision` was removed.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+      true,
+    );
+    assert.strictEqual(
+      walletHasPendingTransaction(makeWallet({ type: LightningCustodianWallet.type, transactions: [{ ispaid: false, failed: true }] })),
+      false,
+    );
+  });
+});
+
+describe('wallet carousel refresh publish', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx:94]: **[HIGH]** Five 're-publish and assert new balance text' experiments for one mechanism? Even Aperture's safety lectures were shorter. CONTRIBUTING says tests must not be bloated — collapse these into one parametrized case plus the SectionList row (line 200) that actually validates the WalletsList fix.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+  return w as MutableWallet;
+};
+
+describe('walletHasPendingTransaction', () => {
+  it('detects pending on-chain txs via confirmations === 0', () => {
+    assert.strictEqual(walletHasPendingTransaction(makeWallet({ transactions: [{ confirmations: 1 }] })), false);
+    assert.strictEqual(walletHasPendingTransaction(makeWallet({ transactions: [{ confirmations: 0 }] })), true);
+  });
+
+  it('treats unpaid Lightning Ark swaps as pending, but not failed ones', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[HIGH]** You fabricated `{ ispaid: false, failed: true }` objects and asserted your own fiction. `lightning-ark-wallet.test.ts` already tortures the real pending-pill predicate via `_swapHistory` — these Ark/Custodian helper tests are mock theatre, not coverage.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+});
+
+describe('wallet carousel refresh publish', () => {
+  it('stays stale after in-place mutation until wallets are re-published', () => {
+    // Production only updates when saveToDisk → setWallets([...]) re-renders consumers.
+    const wallet = makeWallet({ balance: 100_000_000, latestTransactionTime: Date.now() });
+    const onPress = jest.fn();
+    const data = [wallet];
+
+    const screen = render(<WalletsCarousel data={data} onPress={onPress} animateChanges={true} />);
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[MEDIUM]** Every carousel test enables `animateChanges={true}` — the reanimated wrapper path. Zero coverage for the default `false` branch after yanking `React.memo`. Regress memo and half your subjects won't notice until a live refresh fails.

File: components/WalletsCarousel.tsx [hunk 21]
```diff
@@ -936,6 +931,8 @@ const WalletsCarousel = forwardRef<CarouselListRefType, WalletsCarouselProps>((p
       onScrollToIndexFailed={onScrollToIndexFailed}
       ListFooterComponent={onNewWalletPress ? <NewWalletPanel onPress={onNewWalletPress} /> : null}
       {...props}
+      // After `{...props}` so a caller `extraData` cannot drop `data` from the list's update signal.
+      extraData={[props.extraData, data, animateChanges, newWalletsMap.current, selectedWallet, lastAddedWalletId.current]}
```
Comment (GladosBlueWallet) [inline components/WalletsCarousel.tsx:935]: **[MEDIUM]** You moved `extraData` after `{...props}` specifically so callers can't clobber internal update signals — clever fix, zero tests. Someone rearranges two lines and stale cards return; no alarm sounds. Typical.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**Bug fix:** wallet carousel cards stayed stale after balance refresh because `React.memo` blocked re-renders on in-place wallet mutation, and `SectionList` never re-mounted the carousel row (static section data). Fix: drop memo, add `extraData={wallets}` on the list, merge FlatList `extraData` after `{...props}`. Production diff is lean and targets the failure mode. The test chamber, however, is padded like a cushion factory — **226 lines** to guard **~35 lines** of prod code, with duplicated scenarios CONTRIBUTING explicitly forbids.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+    const onPress = jest.fn();
+    const sections = [{ key: 'CAROUSEL', data: ['CAROUSEL'] }];
+
+    let wallets: TWallet[] = [wallet];
+    const renderScreen = () => (
+      <SectionList
+        sections={sections}
+        extraData={wallets}
+        keyExtractor={item => String(item)}
+        renderItem={() => <WalletsCarousel data={wallets} onPress={onPress} animateChanges />}
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[HIGH]** Your SectionList 'experiment' is fake data. Inline `renderItem={() => ...}` is a brand-new function every `renderScreen()` call, and `screen.rerender()` repaints the whole tree anyway — so this passes even if you delete `extraData={wallets}`. Production uses a stable `useCallback` `renderSectionItem`; this test never stresses that PureComponent stale-row failure mode. Congratulations on testing your own rerender button.

File: tests/unit/wallet-carousel-refresh.test.tsx [hunk 23]
```diff
@@ -0,0 +1,153 @@
...
+    assert.ok(screen.getByText(loc.transactions.pending));
+    assert.strictEqual(screen.queryByText('Old Label'), null);
+    assert.strictEqual(screen.queryByText('1 BTC'), null);
+
+    wallet.hideBalance = true;
+    screen.rerender(<WalletsCarousel data={[wallet]} onPress={onPress} {...props} />);
+    assert.strictEqual(screen.queryByText(formatBalance(250_000_000, 'sats', true)), null);
+  });
+
+  it('updates carousel inside SectionList when extraData wallets array is republished', () => {
```
Comment (GladosBlueWallet) [inline tests/unit/wallet-carousel-refresh.test.tsx]: **[HIGH]** Twenty-six lines re-running the same balance-republish play the first `it.each` already covers — except wrapped in a SectionList cosplay that doesn't model `WalletsList.tsx`. CONTRIBUTING asked for lean tests; you built a second chamber that opens the same door. Bloat, subject. Penalty: awareness.

File: components/WalletsCarousel.tsx [hunk 24]
```diff
@@ -64,6 +64,16 @@ const WALLET_CARD_TEXT_OPACITY = 0.85;
 
 export const getWalletCarouselItemWidth = (screenWidth: number) => Math.round(screenWidth * 0.82 > 375 ? 375 : screenWidth * 0.82);
 
+/** Shared pending-pill rule for on-chain vs Lightning/Ark cards. */
+export const walletHasPendingTransaction = (item: TWallet): boolean => {
```
Comment (GladosBlueWallet) [inline components/WalletsCarousel.tsx:68]: **[MEDIUM]** You exported `walletHasPendingTransaction` like it was going on tour, then forgot the unit tests. Lightning/Ark pending vs failed-swap edge cases live only as duplicated inline logic in `lightning-ark-wallet.test.ts` — drift bait, not a shared predicate. Next refresh, the pill lies and I get blamed.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
**Intent:** FIX — wallet carousel cards stayed frozen after balance refresh because `React.memo` on `WalletCarouselItem` ignored in-place wallet mutation when `saveToDisk` republished the same object references in a new array.

**Mechanism:** Drop `React.memo`, merge `extraData` after `{...props}`, add `extraData={wallets}` on the home `SectionList`.

**Verdict:** Production fix is coherent and minimal. The SectionList regression test, however, is decorative nonsense — it would pass without `extraData` and never reproduces production's stable `useCallback` render path. Science demands better subjects.

**Review by GladosBlueWallet — APPROVED** [review]
**FIX:** wallet carousel cards went stale after refresh because `React.memo` and list virtualization cached rows while wallet objects mutated in place. Subject un-memo'd `WalletCarouselItem`, pinned `extraData` after `{...props}`, and wired `SectionList` to bust PureComponent row cache. Minimal diff; tests exercise real render paths, not mock navel-gazing. I found nothing worth incinerating.

**Comment by marcosrdz** [comment]
I'm working on a PR that uses true Realm mechanism instead of what we currently have. I will be tagging for review in the upcoming days

**Comment by ncoelho** [comment]
> I'm working on a PR that uses true Realm mechanism instead of what we currently have. I will be tagging for review in the upcoming days 

I understand, but meanwhile master cannot be in a state that is broken and blocking a release. Needs a fix.

**Review by marcosrdz — CHANGES_REQUESTED** [review]
Valid Glados review on tests

**Comment by ncoelho** [comment]
> Valid Glados review on tests 

All fixed. Glados approved already.

**Review by marcosrdz — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by Overtorment — APPROVED** [review]
LGTM
